packages/extensions/openapi-to-typespec/src/generate/generate-arm-resource.ts

```diff
@@ -25,9 +26,48 @@ export function generateArmResource(resource: TspArmResource): string {
     definitions.push(`@parentResource(${resource.resourceParent.name})`);
   }
 
-  definitions.push(`model ${resource.name} is ${resource.resourceKind}<${resource.propertiesModelName}> {`);
+  if (
+    getArmCommonTypeVersion() &&
+    !resource.propertiesPropertyRequired &&
+    resource.propertiesPropertyVisibility.length === 2 &&
+    resource.propertiesPropertyVisibility.includes("read") &&
+    resource.propertiesPropertyVisibility.includes("create")
+  ) {
+    definitions.push(`model ${resource.name} is ${resource.resourceKind}<${resource.propertiesModelName}> {`);
 
-  definitions = [...definitions, ...getModelPropertiesDeclarations(resource.properties)];
+    definitions = [...definitions, ...getModelPropertiesDeclarations(resource.properties)];
+  } else {
+    definitions.push(
+      `#suppress "@azure-tools/typespec-azure-core/composition-over-inheritance" "For backward compatibility"`,
+    );
+    definitions.push(
+      `#suppress "@azure-tools/typespec-azure-resource-manager/arm-resource-invalid-envelope-property" "For backward compatibility"`,
+    );
+    definitions.push(`@includeInapplicableMetadataInPayload(false)`);
+
+    if (!getArmCommonTypeVersion()) {
+      if (resource.baseModelName) {
+        definitions.push(`model ${resource.name} extends ${resource.baseModelName} {`);
+      } else {
```

Comment:
I wonder if we could abstract all of this new logic into a function and just call the function here to help readability and maintenance in the future.